src/backend/distributed/worker/worker_partition_protocol.c

```diff
@@ -1156,29 +1253,56 @@ RangePartitionId(Datum partitionValue, const void *context)
 /*
  * HashPartitionId determines the partition number for the given data value
  * using hash partitioning. More specifically, the function returns zero if the
- * given data value is null. If not, the function applies the standard Postgres
- * hashing function for the given data type, and mods the hashed result with the
- * number of partitions. The function then returns the modded number as the
- * partition number.
+ * given data value is null. If not, the function follows the exact same approach
+ * as Citus distributed planner uses.
  *
- * Note that any changes to PostgreSQL's hashing functions will reshuffle the
- * entire distribution created by this function. For a discussion of this issue,
- * see Google "PL/Proxy Users: Hash Functions Have Changed in PostgreSQL 8.4."
+ * Note that this function also supports deprecated APIs, which uses a slightly
+ * different hash calculation method.
  */
 static uint32
 HashPartitionId(Datum partitionValue, const void *context)
 {
 	HashPartitionContext *hashPartitionContext = (HashPartitionContext *) context;
 	FmgrInfo *hashFunction = hashPartitionContext->hashFunction;
 	uint32 partitionCount = hashPartitionContext->partitionCount;
-	Datum hashDatum = 0;
-	uint32 hashResult = 0;
+	ShardInterval **syntheticShardIntervalArray =
+		hashPartitionContext->syntheticShardIntervalArray;
+	FmgrInfo *comparisonFunction = hashPartitionContext->comparisonFunction;
+	Datum hashDatum = FunctionCall1(hashFunction, partitionValue);
+	int32 hashResult = 0;
 	uint32 hashPartitionId = 0;
 
-	/* hash functions return unsigned 32-bit integers */
-	hashDatum = FunctionCall1(hashFunction, partitionValue);
-	hashResult = DatumGetUInt32(hashDatum);
-	hashPartitionId = (hashResult % partitionCount);
+	/*
+	 * This is to provide backward compatibility. See comments in
+	 * worker_hash_partition_table().
+	 */
+	if (hashPartitionContext->deprecatedAPIcall)
```

Comment:
it might be cleaner to have an alternative implementation of `HashPartitionId` that we pass to `FilterAndPartitionTable` when using the deprecated API, which seems acceptable since it's only 10 lines of code.